File: dart/lib/src/record/get_record_v1_address.dart
import '../constants/records.dart';
import '../domain/get_domain_address.dart';

/// Parameters for getting a V1 record address
class GetRecordV1AddressParams {
  const GetRecordV1AddressParams({
    required this.domain,
    required this.record,
  });

  /// The domain under which the record resides
  final String domain;

  /// The type of record to derive the address for
  final Record record;
}

/// Derives the address of a version 1 record.
///
/// This function mirrors js-kit/src/record/getRecordV1Address.ts
///
/// [params] - Parameters containing the domain and record type
///
/// Returns the derived record address
Future<String> getRecordV1Address(GetRecordV1AddressParams params) async {
  final result = await getDomainAddress(GetDomainAddressParams(
    domain: '${params.record.name}.${params.domain}',
    record: RecordVersion.v1,
  ));

  return result.domainAddress;
}

File: js-kit/src/record/getRecordV1Address.ts
import { getSnsDomainAddress } from "../domain/getSnsDomainAddress";
import { Record, RecordVersion } from "../types/record";

/**
 * Parameters for deriving a V1 record address.
 *
 * @example
 * ```ts
 * const params: GetRecordV1AddressParams = { domain: "example", record: Record.Url };
 * ```
 */
export interface GetRecordV1AddressParams {
  /** TLD-less domain name. */
  domain: string;
  /** Record type. */
  record: Record;
}

/**
 * Derives the address of a V1 record account.
 *
 * The V1 account is derived by prefixing the record label to the domain name.
 *
 * @param params Record address derivation parameters
 * @param params.domain TLD-trimmed SNS domain name
 * @param params.record Record type
 * @returns The derived V1 record account address.
 *
 * @example
 * ```ts
 * const address = await getRecordV1Address({ domain: "example", record: Record.Url });
 * ```
 */
export const getRecordV1Address = async ({
  domain,
  record,
}: GetRecordV1AddressParams) => {
  const { domainAddress } = await getSnsDomainAddress({
    domain: record + "." + domain,
    record: RecordVersion.V1,
  });

  return domainAddress;
};
